compiler/src/iree/compiler/Codegen/TransformDialectExtensions/TransformDialectExtensions.cpp

```diff
@@ -59,6 +63,111 @@ static SmallVector<int64_t> getIndicesOfDynamicDims(ShapedType t) {
 // Patterns for ForeachThreadOpToFlow rewrite.
 //===---------------------------------------------------------------------===//
 
+/// Iteratively populate the computations that define `foreachThreadOp`'s
+/// getNumThreads into the workgroup_count region of `dispatchOp` and set the
+/// `dispatchOp` workload operands to the values required above.
+/// Assumes the Flow::DispatchWorkgroupsOp is built with an empty region.
+static void populateWorkgroupCountComputingRegion(
+    PatternRewriter &rewriter, scf::ForeachThreadOp foreachThreadOp,
+    Flow::DispatchWorkgroupsOp dispatchOp) {
+  Location loc = foreachThreadOp.getLoc();
+  OpBuilder::InsertionGuard g(rewriter);
+  Region &r = dispatchOp.workgroup_count();
+  assert(r.empty() && "expected block-less workgroup_count region");
+  Block *block = rewriter.createBlock(&r);
+  rewriter.setInsertionPointToStart(block);
+  SmallVector<Value> workgroupCount{foreachThreadOp.getNumThreads()};
+  // Resize to `3` to match IREE's assumptions.
+  workgroupCount.resize(3, rewriter.create<arith::ConstantIndexOp>(loc, 1));
+  auto returnOp = rewriter.create<Flow::ReturnOp>(loc, workgroupCount);
+
+  // Start filling the region.
+  // Create a worklist of the Operations creating the values returned.
+  // Iteratively pull the first operation that is not dominated by any others
+  // and add the operations defining it to the worklist unless either:
+  //   1. the operation consumes any tensor, we avoid pulling tensor arguments
+  //      into the workgroup count region.
+  //   2. the operation is null (i.e. the operand is a block argument), in which
+  //      case we just pass the block argument without trying to analyze
+  //      further.
+  //   3. the operation has already been seen, in which case we avoid
+  //      reprocessing.
+  {
```

Comment:
Not sure why we need to do this. How do we avoid operations that are compute operations being pulled into the workgroup count region?

Reply:
Well that is the point .. if we need to bring in `affine.min` and other `compute_tile_size_helper_fun()` to compute the actual sizes then we would. This would work with `tensor.dim` operands and specifically avoids pulling in anything non-index. The result is "n" index parameters (the "workload") and 3 index results (the "workgroup_count").
"n" is dependent on the logic in the IR that is used to compute the sizes and there is nothing hardcoded.